diff --git a/gum/gum.c b/gum/gum.c
index 58e83a943b603c1fb9a66196f5b7323c9701396f..f1dbdb68fc622593b26db121bc5780d3626642dc 100644
--- a/gum/gum.c
+++ b/gum/gum.c
@@ -302,7 +302,7 @@ gum_init_embedded (void)
   g_log_set_default_handler (gum_on_log_message, NULL);
   gum_do_init ();
 
-  g_set_prgname ("frida");
+  g_set_prgname ("gogogo");
 
 #if defined (HAVE_LINUX) && defined (HAVE_GLIBC)
   gum_libdl_prevent_unload ();
